### DISCUSSION TITLE
3D interpolation for unordered points - performance

### DISCUSSION THREAD

**Discussion by mcacace** [idaholab/moose#17878] [Q&A General] [answered]
Dear all,
I am facing some performance issues while trying to implement a function to interpolate a scalar value from 3D unordered sampling points to set up an initial condition for a primary variable. The basic structure is simple. 
I am reading upon the constructor call of the function the sampling points and store their values in vectors (I found they are faster than unordered maps). Based on that I do perform the interpolation at the call of the value() function.  This is where my pain starts. Given that I have no topological ordering in the sample and target data, I have to rely on a global interpolation - currently this is done as a global search weighted by the distance of the sampling points to the targeted one restricted over a searching sphere of a given radius and provides (really!) terrible scalability. This comes with no surprise (looping through global search is anything but clever), but the performance is way lower than I could foresee. As an example, running it over 2 millions of dofs and for a similar size of the sampling points, the initialization phase runs for more than 3 hours on a cluster. I have been looking at the framework but it does seem that all interpolation routines available are either based on  2D cases, or required some degree of ordering for the sampling points (though I might be wrong here). 
Before trying to port a 3D NN interpolation algorithm I have been using in an external application to the framework (which would require extensive coding), I was wondering if anyone could point me to a possible (easier and faster) solution.
As always, any help is highly appreciated. 
Thanks, Mauro

**Comment by ABallisat** [comment]
Take a look at a k-d tree for some very fast nearest neighbour lookup. I have been using the SciPy implementation to do high dimensional interpolation with millions of points using inverse distance weighted interpolation in Python and it is very quick to do on a standard laptop. You can probably dig into the source code of that as an example of how to implement it.

**Reply by mcacace** [reply]
@ABallisat, thanks for the hint. I see that there is some implementation of kd-tree and neighbour search in libmesh. Will give it a try and see whether it could help. Again, thanks

**Comment by lynnmunday** [comment]
Could you try something like what @hugary1995 suggested in this discussion: https://github.com/idaholab/moose/discussions/17806#discussioncomment-745928
we're going to try to use this method to do something similar to what you're doing.  If you had your point value data in csv format, you could read it in with a csv value postprocessor and then interpolate it to the nodes with the reaction kernel.
I would be really interested in finding out if this would work for you because we've only tried this with small problems.  If this is really slow we'll need to find ways to speed it up.

**Reply by mcacace** [reply]
@lynnmunday, not sure I did understand all from the discussion you posted (lost myself in trying to relate an IC action to a Dirac Kernel action, and how to organize the reading and passing of information from a post processor that reads an external file to initialize a variable - late hours here in Berlin), but will give it some more thoughts in the next days. Thanks for your help

**Reply by lynnmunday** [reply]
You're right, the dirac kernel would not work for IC's.  I got that all wrong.

**Comment by mcacace** [comment, marked as answer]
Dear all,
I would like to thanks again for your hints. I implemented an interpolation based on a kdtree neighbour search and for what I tested so far it works nice - 2 millions points runs over few seconds on my laptop. I will do some more tests (by increasing the sampling points dimensions) to see whether all works fine (not sure if there is a limit size for initializing the kdtree), but I would now consider this post as closed.

**Reply by ABallisat** [reply]
glad it worked, could you report back on your findings so we know if it does scale or not and what the limits are please. Thanks in advance!

**Reply by mcacace** [reply]
Just checked on a real case examples. sampling points size is something more than 2 millions points, and target points size almost the same. It runs within few second. Attached a picture of the final results - the (close not to be seen points are the sampling points on top of the interpolated values. This said, I do think that the kd-tree has no memory limits for realistic applications... 
![screen](https://user-images.githubusercontent.com/13695024/119153828-deeb1e00-ba51-11eb-89cf-7fc6d08ece99.png)
